File: scope104/src/main/java/com/petro/scope104/data/UserDataRepository.java
package com.petro.scope104.data;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.petro.scope104.data.qualifier.DataBaseQualifier;
import com.petro.scope104.data.qualifier.Network;
import com.petro.scope104.domain.entity.WorkerEntity;
import com.petro.scope104.presentation.list.Gender;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;

public class UserDataRepository {
    private final UserDataSource dataBaseDataSource;
    private final UserDataSource networkDataSource;

    @Inject
    public UserDataRepository
            (@DataBaseQualifier UserDataSource dataBaseDataSource,
             @Network UserDataSource networkDataSource
    ) {
        this.dataBaseDataSource = dataBaseDataSource;
        this.networkDataSource = networkDataSource;
    }

    public Observable<List<WorkerEntity>> loadUsers(int pageNumber, int pageSize, @NonNull Gender gender, @NonNull List<String> countries) {
        final Observable<List<WorkerEntity>> dbLiveData = dataBaseDataSource.loadUsers(pageNumber, pageSize, gender, countries);
        return dbLiveData.switchMap(dbInput -> {
            if (dbInput.size() >= pageSize) {
                return Observable.just(dbInput);
            } else {
                final Observable<List<WorkerEntity>> networkLiveData = networkDataSource.loadUsers(pageNumber, pageSize, gender, countries);
                return networkLiveData.map(networkInput -> {
                    dataBaseDataSource.saveUsers(networkInput);
                    return networkInput;
                });
            }
        });
    }
}

File: scope104/src/main/kotlin/com/petro/scope104/data/UserDataRepository.kt
package com.petro.scope104.data

import com.petro.scope104.data.qualifier.DataBaseQualifier
import com.petro.scope104.data.qualifier.Network
import com.petro.scope104.domain.entity.WorkerEntity
import com.petro.scope104.presentation.list.Gender
import io.reactivex.rxjava3.core.Observable
import javax.inject.Inject

class UserDataRepository @Inject constructor(
    @param:DataBaseQualifier private val dataBaseDataSource: UserDataSource,
    @param:Network private val networkDataSource: UserDataSource
) {
    fun loadUsers(
        pageNumber: Int,
        pageSize: Int,
        gender: Gender,
        countries: List<String>
    ): Observable<List<WorkerEntity>> {
        val dbLiveData: Observable<List<WorkerEntity>> =
            dataBaseDataSource.loadUsers(pageNumber, pageSize, gender, countries)
        return dbLiveData.switchMap { dbInput: List<WorkerEntity> ->
            if (dbInput.size >= pageSize) {
                return@switchMap Observable.just<List<WorkerEntity>>(dbInput)
            } else {
                val networkLiveData: Observable<List<WorkerEntity>> =
                    networkDataSource.loadUsers(pageNumber, pageSize, gender, countries)
                return@switchMap networkLiveData.map<List<WorkerEntity>> { networkInput: List<WorkerEntity> ->
                    dataBaseDataSource.saveUsers(networkInput)
                    networkInput
                }
            }
        }
    }
}